dcplib/security/authn.py

```diff
@@ -0,0 +1,114 @@
+gserviceaccount_domain = "iam.gserviceaccount.com"
+
+# recycling the same session for all requests.
+session = requests.Session()
+
+
+@functools.lru_cache(maxsize=32)
+def get_openid_config(openid_provider: str):
+    """
+
+    :param openid_provider: the openid provider's domain.
+    :return: the openid configuration
+    """
+    if openid_provider.endswith(gserviceaccount_domain):
+        openid_provider = 'accounts.google.com'
+    elif openid_provider.startswith("https://"):
+        openid_provider = furl(openid_provider).host
+    res = requests.get("https://{op}/.well-known/openid-configuration".format(op=openid_provider))
+    res.raise_for_status()
+    return res.json()
+
+
+def get_jwks_uri(openid_provider):
+    if openid_provider.endswith(gserviceaccount_domain):
+        return "https://www.googleapis.com/service_accounts/v1/jwk/{op}".format(op=openid_provider)
+    else:
+        return get_openid_config(openid_provider)["jwks_uri"]
+
+
+@functools.lru_cache(maxsize=32)
+def get_public_keys(openid_provider: str):
+    """
+    Fetches the public key from an OIDC Identity provider to verify the JWT.
+    :param openid_provider: the openid provider's domain.
+    :return: Public Keys
+    """
+    keys = session.get(get_jwks_uri(openid_provider)).json()["keys"]
+    return {
+        key["kid"]: rsa.RSAPublicNumbers(
+            e=int.from_bytes(base64.urlsafe_b64decode(key["e"] + "==="), byteorder="big"),
+            n=int.from_bytes(base64.urlsafe_b64decode(key["n"] + "==="), byteorder="big")
+        ).public_key(backend=default_backend())
+        for key in keys
+    }
+
+
+def decode_jwt(token: str, audience: typing.List[str]) -> typing.Optional[typing.Mapping]:
+    """
+    Verify and decode the JWT from the request.
+
+    :param token: the Authorization header in the request.
+    :param audience: the expected audience for the JWT
+    :return: Decoded and verified token.
+    """
+    try:
+        unverified_token = jwt.decode(token, verify=False)
+    except jwt.DecodeError:
+        raise AuthenticationException('Failed to decode token.')
+
+    assert_authorized_issuer(unverified_token)
+    issuer = unverified_token['iss']
+    public_keys = get_public_keys(issuer)
+
+    try:
+        token_header = jwt.get_unverified_header(token)
+        verified_tok = jwt.decode(token,
+                                  key=public_keys[token_header["kid"]],
+                                  issuer=issuer,
+                                  audience=audience,
+                                  algorithms=allowed_algorithms,
+                                  )
+    except jwt.PyJWTError as ex:  # type: ignore
+        raise AuthenticationException('Authorization token is invalid') from ex
+    return verified_tok
+
+
+def assert_authorized_issuer(token: typing.Mapping[str, typing.Any]) -> None:
+    """
+    Must be either `Config.get_openid_provider()` or in `Config.get_trusted_google_projects()`
+    :param token: dict
```

Comment:
nit: `dict` duplicated the `token`'s typing hint above.

Reply:
changing to "a JWT from the authorization header of a request"